### PR TITLE
feature: Add support for custom method placement in `ordered_class_elements`

### LINKED ISSUE #6252: Add support of Codeception ordering (_before, _after, etc.)
## Feature request

Add support of Codeception - PHP Testing framework.

Unlike the phpunit, instead of "setUp" and "tearDown" the "_before" and "_after" method are used here. 
 [Codeception documentation](https://codeception.com/docs/reference/Module.html#algolia:p:nth-of-type(28))

So this feature should add "codeception" special type in OrderedClassElementsFixer for correct sorting these methods 
'_before',
'_beforeStep',
'_beforeSuite',
'_after',
'_afterStep',
'_afterSuite'

### PR DESCRIPTION
**Pull request by bakerkretzmar**
As discussed in #6244, this PR adds the ability to place specific individual methods in a custom order within a class.

The original PR was to add a custom `invoke` option similar to `construct`, to place the `__invoke()` method at a custom location instead of grouping it with other magic methods. After some discussion with @SpacePossum @gharlan and @driftingly it sounded like a more generic solution would be preferred, to allow placing any specific method at a custom location, not just `__invoke()`.

The suggestion from that discussion was to allow `['method' => 'methodName']` in the list of options to specify custom orders. I chose to use `method:methodName` instead because the code changes are a lot simpler and smaller if all the options are still plain strings.

`method:` options override any other ordering, so for example `method:__invoke` takes precedence over `magic`.

In the future it would probably be quite straightforward to support specific properties/constants/traits too with similar syntax (e.g. `property:propertyName`), although I'm not sure if that's useful.

Replaces #6244
Closes #6252

### CODE REVIEW

**Comment by bakerkretzmar** [comment]
@keradus @SpacePossum @gharlan anything else you need from me here?

**Comment by driftingly** [comment]
Thanks, @bakerkretzmar for updating this!

**Comment by Wirone** [comment]
@bakerkretzmar I wanted to rebase it in order to process it further on latest `master`, but it seems like your changes in docs were made by hand, not with `dev-tools/doc.php` (how checks have passed 🤔?). There's no way to achieve "subset of" allowed values with callable, which is unfortunate. It would be great to find a way to document `method:*` in user-friendly manner. I really would like to get this merged, I really like the idea 🙂.

**Comment by bakerkretzmar** [comment]
@Wirone thanks! I did update the docs by hand, I had no idea `dev-tools/doc.php` existed 😄 I'm not sure how to document `method:*` automatically... I based the closure in this PR on some others I found in the codebase like https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/c1470f11ff899ad2f907a1ed5e3e17ea4dbea53b/src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php#L209, https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/c1470f11ff899ad2f907a1ed5e3e17ea4dbea53b/src/Fixer/PhpUnit/PhpUnitTestCaseStaticMethodCallsFixer.php#L373, https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/c1470f11ff899ad2f907a1ed5e3e17ea4dbea53b/src/Fixer/Import/OrderedImportsFixer.php#L260, etc. How are those ones documented?

**Comment by Wirone** [comment]
> I had no idea `dev-tools/doc.php` existed 😄

I don't know if it existed back then, so no worry. That's why I try to introduce some [common scripts](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6839) and improve contributing guide, because I see that not everything is obvious (for me too) 🙂.

In terms of allowed values probably we need to include info about `method:*` in the fixer's description ("List of strings defining order of elements.") or add it to `$specialTypes` so it's automatically generated along with other values? I honestly don't know, I still learn 😅. @kubawerlos can you suggest something?

**Comment by kubawerlos** [comment]
Before we think how, let's discuss if we should do it at all, see https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3864#issuecomment-402388413.

**Comment by kubawerlos** [comment]
Closing/reopening to trigger CI.

**Comment by Wirone** [comment]
@kubawerlos I don't see negative sides of this approach (besides slightly higher complexity to maintain). There's no any "injecting code", only fixed implementation that allows better flexibility of how users want to order methods in a class. It's option like the others, but supports dynamic values - however those values have to be in subset defined by class itself. I believe there are valid scenarios for that 🙂.

**Comment by kubawerlos** [comment]
@Wirone oh, I see now, I understood the docs part:

> a single method name (e.g. ``method:__invoke)`` to set the relative order of that particular method

as there would be possible to use that method to sort class elements, assuming the doc was generated from code, which apparently is not the case here.

There is no way to achieve that now - it's either a fixed list of strings or any string.

One way to go would be to deprecate the current option and create a new one, where all current options would have `@` at the beginning or maybe something smarter, not to confuse with PHPDoc tags, let's say `#` for a moment, and other entries would be interpreted as methods, so config can look like:
```php
[
    '#method_protected_static',
    'MyClass::myMethod',
    '#method_protected',
    '#magic'
]
```

[SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle) for the win - let's split it anyway and add `magic` (and `phpunit`, but I have no clue what it suppose to do) with proper tests for each new type.

**Comment by bakerkretzmar** [comment]
@kubawerlos that's an interesting idea! I think the suggestion of expanding this to properties came up in #6244 as well.

I'd like to keep the scope of this PR very focused, so I'm not going to update it to deprecate the current options or to handle properties in addition to methods—those are good ideas, but my proposal here is **only** to add a new `method:` prefix to adjust the order of specific methods.

@Wirone can this particular fixer just be documented by hand?

Let me know if there's anything I need to do here!

**Comment by Wirone** [comment]
> can this particular fixer just be documented by hand?

No, because it will be overridden by script when it' used. We need to use description for that, so script would take it always the same way.

**Comment by bakerkretzmar** [comment]
@Wirone thanks! I moved things into the description, the documentation script should work now.

File: tests/Fixer/ClassNotation/OrderedClassElementsFixerTest.php
Comment (Wirone) [inline tests/Fixer/ClassNotation/OrderedClassElementsFixerTest.php:1]: We should have test scenarios where:

- `method:*` is used, but specified method does not exist in the class (what should happen? I believe it should be ignored and all methods should be ordered based on rest of the configured options)
- the same `method:*` is used twice (there should be an error pointing to misconfiguration, because the same method can't exist in multiple places at the same time)
- `method:*` contains unsupported chars, like `method:boom()` (misconfiguration error)

File: src/Fixer/ClassNotation/OrderedClassElementsFixer.php
Comment (Wirone) [inline src/Fixer/ClassNotation/OrderedClassElementsFixer.php:228]: I think that `relative` is misleading here. Relative to what? Maybe `explicit` would be better?

**Review by Wirone — CHANGES_REQUESTED** [review]

File: src/Fixer/ClassNotation/OrderedClassElementsFixer.php
Comment (driftingly) [inline src/Fixer/ClassNotation/OrderedClassElementsFixer.php:228]:
```suggestion
- `method:*`: specify a single method name (e.g. `method:__invoke`) to set the explicit order of that specific method'
```

Comment (Wirone) [reply src/Fixer/ClassNotation/OrderedClassElementsFixer.php:228]: Thanks @driftingly for a review, but [I've already pointed this](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6360#discussion_r1192831994) 😉

Comment (driftingly) [reply src/Fixer/ClassNotation/OrderedClassElementsFixer.php:228]: Just easier for @bakerkretzmar to include the suggestion 😄

File: tests/Fixer/ClassNotation/OrderedClassElementsFixerTest.php
Comment (bakerkretzmar) [reply tests/Fixer/ClassNotation/OrderedClassElementsFixerTest.php:1]: Great ideas, will do!

File: src/Fixer/ClassNotation/OrderedClassElementsFixer.php
Comment (bakerkretzmar) [reply src/Fixer/ClassNotation/OrderedClassElementsFixer.php:228]: > Relative to what?

Relative to the other class elements. Using something like `method:__invoke` doesn't set the explicit order of the `__invoke` method (e.g. 2nd or 3rd element in the class), it sets the order _relative to the other elements_ configured in this rule (e.g. _before_ or _after_ some other elements).

Not super attached to it though, I can just take that word out and I think it's still clear.

Comment (Wirone) [reply src/Fixer/ClassNotation/OrderedClassElementsFixer.php:228]: Ok, so maybe we can use this instead?

```suggestion
- `method:*`: specify a single method name (e.g. `method:__invoke`) to set the explicit placement of that specific method'
```

File: tests/Fixer/ClassNotation/OrderedClassElementsFixerTest.php
Comment (bakerkretzmar) [reply tests/Fixer/ClassNotation/OrderedClassElementsFixerTest.php:1]: @Wirone I added a test case with a missing method name where it just doesn't do anything.

I like the other two ideas but I think they might out of scope for this PR. I tested the current behaviour of both of those cases and neither is an error right now, and I think the way they both work makes sense:

- If the same config element appears twice, the _last_ time it appears is the one that takes precedence. This is how it already works if you repeat e.g. `method_public`. An error might make more sense here but for now this PR's behaviour is consistent with the existing behaviour.
- If a config element contains unsupported characters it's ignored. E.g. if you include `magic()` or `construct()` it just doesn't do anything. Technically this makes some sense since there is not a method called `construct()`, so it's similar to using a method name that doesn't exist.. An error makes a lot more sense here too, but right now this PR matches the existing behaviour too.

Thoughts?

File: src/Fixer/ClassNotation/OrderedClassElementsFixer.php
Comment (bakerkretzmar) [reply src/Fixer/ClassNotation/OrderedClassElementsFixer.php]: Sorry only just saw this comment now—I took out "relative", what do you think? Now it just says "to set the order of that specific method".

File: tests/Fixer/ClassNotation/OrderedClassElementsFixerTest.php
Comment (Wirone) [reply tests/Fixer/ClassNotation/OrderedClassElementsFixerTest.php:1]: LGTM 👍

File: src/Fixer/ClassNotation/OrderedClassElementsFixer.php
Comment (Wirone) [reply src/Fixer/ClassNotation/OrderedClassElementsFixer.php]: It's OK with current description ("to set the order of that specific method"), so it's up to you what you like better - I'll just resolve this thread and you may apply suggestion as you wish 🙂.

Comment (bakerkretzmar) [reply src/Fixer/ClassNotation/OrderedClassElementsFixer.php]: Thanks! I'm going to keep what's there now, just "set the order", but I appreciate the input. And of course feel free to change it yourself later if you think of something better!

**Review by Wirone — APPROVED** [review]
👍 from me, good job @bakerkretzmar. I won't merge it, though, let's wait for another review.

**Review by kubawerlos — APPROVED** [review]

**Comment by Wirone** [comment]
Thank you @bakerkretzmar and welcome to the Fixer's contributors group 😁🍻.

**Comment by bakerkretzmar** [comment]
Thanks!! Really appreciate all the feedback on this and glad we're going to be able to use it soon 🔨 🔧

**Comment by ldommer** [comment]
Hey there, great idea for a new `ordered_class_elements` fixer config and great that it has been added to the package.

Recently we discussed (opinionated) class method sorting in our team during a tech talk and I wrote a preset to our (Laravel Pint) config file, that matches some of our needs.

We do not only want to sort class methods based on their type (static vs. non-static) and visibility, but also regarding their meaning within framework functionality (Laravel in our case, but I think this is also relevant to all other PHP frameworks).

This means we also want to sort class methods by groups like `Model Events / Lifecycle`, `Query Scopes`, `Relations` and `Accessors / Mutators`. By now we have to do it manually, but I think it would be a nice addition to this fixer and config.

The following sortings could be achieved with only two small additions to this fixer and config:

- `Model Events / Lifecycle`: have specific method names (requires method name comparison, works)
- `Query Scopes`: have unspecific method names that always start with the string `scope` (requires method name `starts with string comparison` or `wildcard matching`, does not work)
- `Relations`: have unspecific return types that always start with a string of a fixed list of strings (requires return type `starts with string comparison` or `wildcard matching`, does not work)
- `Accessors / Mutators`: have specific return types (requires return type comparison, does not work)

Is there any chance that the two necessary additions (1. method name `starts with string comparison` or `wildcard matching` and 2. `return type comparison` and `starts with string comparison` or `wildcard matching`) will be added in the future or might be merged if I fork the repo and develop it by myself and PR it?

**Pull request merged**
